modules/internal/ChapelHashtable.chpl

```diff
@@ -356,6 +356,121 @@ module ChapelHashtable {
+    // table until they hit 'a' full slots, and then yield the next
+    // (c - a + 1) full slots.
+    // This should perform in roughly linear time unless the slots
+    // in the table are poorly distributed or there are a lot of
+    // deleted slots in the table.
+    // Performance will also degrade if a different leader is used
+    // that creates a lot of smaller ranges over the iteration
+    // space.
+
+    iter allFilledSlots() ref {
+      for i in 0..#tableSize do
+        if table[i].isFull() then yield table[i];
+    }
+
+    iter allFilledSlots(param tag: iterKind) ref
+      where tag == iterKind.standalone {
+      
+      if debugDefaultAssoc then
+        writeln('*** In chpl__hashtable.allFilledSlots ' +
+                'standalone iterator');
+
+      const size = tableNumFullSlots;
+      const numChunks = _allSlotsNumChunks(size);
+
+      if debugDefaultAssoc then
+        writeln('Iterating over ', size, ' elements across ',
+                numChunks, ' chunks');
+
+      if numChunks == 1 {
+        for i in 0..#tableSize do
+          if table[i].isFull() then yield table[i];
+      } else {
+        coforall chunk in 0..#numChunks {
+          const (lo, hi) = _computeBlock(size, numChunks, chunk,
+                                         size-1);
+          const req = hi - lo + 1;
+          var skip = 0;
+          var count = 0;
+
+          if req > 0 then for i in 0..#tableSize {
+            ref slot = table[i];
+
+            // Iterate through table until we hit 'a' full slots.
+            if skip < lo {
+              if slot.isFull() then skip += 1;
+              continue;
+            }
+
+            // Yield 'req' full slots.
+            if count >= req then break;
+            if slot.isFull() {
+              count += 1;
+              yield slot;
+            }
+          }
+
+          if debugAssocDataPar then
+            writeln('Chunk ', chunk, 'required ', req,
+                    ' and yielded ', count, ' full slots from ',
+                    lo, ' to ', hi);
+        }
+      }
+    }
+
+    iter allFilledSlots(param tag: iterKind)
+      where tag == iterKind.leader {
+
+      if debugDefaultAssoc then
+        writeln('*** In chpl__hashtable.allFilledSlots ' +
+                'leader iterator');
+
+      // Let the range leader break the space up into even chunks.
+      var space = 0..#tableNumFullSlots;
+      for followThis in space.these(tag) do yield followThis;
+    }
+
+    iter allFilledSlots(param tag: iterKind, followThis) ref
+      where tag == iterKind.follower {
+
+      if debugDefaultAssoc then
+        writeln('*** In chpl__hashtable.allFilledSlots ' +
+                'follower iterator');
+
+      const (lo, hi) = followThis;
+      const req = hi - lo + 1;
+      var skip = 0;
+      var count = 0;
+
+      if req > 0 then for i in 0..#tableSize {
+        ref slot = table[i];
+
+        // Iterate through table until we hit 'a' full slots.
```

Comment:
This doesn't have perfect speedup because each task will read the start of the table.
If this is the iteration strategy we end up with, I think we need a different way to compute it.

Reply:
I was looking at this PR in context of https://github.com/Cray/chapel-private/issues/1012 (looking at this PR for API help and seeing if I should just use Set instead of using ChapelHashtable directly)

I'm not totally sure I grok these changes, but it seems like we're trying to update the default associative arrays to work better in a zippered context. I think that's worthwhile, but I'd rather see it in a different PR, especially so it's easier to look at the performance impact.

Reply:
The idea proposed by @cassella is to support zippering.

The problem is that even between two different hashtables the distribution of keys might be such that one task yields next to no slots (I suppose you could make the argument "just have a good hashing function, then", which is fair.)

I will go ahead and revert to calling `allSlots` in this effort and open a separate PR with `allFilledSlotsByRef` (or whatever we name it).